### DISCUSSION TITLE
Document duplication #152

### DISCUSSION THREAD

**Discussion by wancom** [wonder-wonder/cakemix-server#232] [Design]
As issued #152, we will add new feature for document duplication.
This is especially useful for template support.

**Comment by wancom** [comment]
My current idea is as follows.

The new API duplicates the latest document into specified folder.
The permission and other attributes are not succeed and applied the current user.

New API endpoint
`/doc/:folderid/:docid`

`folderid` is folder where copied document is saved.
`docid` is original document.

FYI:
Document creation API is `/doc/:folderid`.
The  suggested API is combined creation API and `/:docid`

**Reply by 0eta0** [reply]
I want to know about the permission, for example, can I use this API for a document that I only have read permission?

**Reply by wancom** [reply]
> can I use this API for a document that I only have read permission?

Yes, you can.

The API only requires the read permission for original document (and write permission to create new document.)
The base concept is just copy and paste so the read permission for original document is enough.

**Reply by 0eta0** [reply]
Thank you!

**Reply by wancom** [reply]
The API endpoint looks OK, but it would conflict in future (like /doc/:docid/somethingnew and /doc/:folderid/:docid)

I suggest another idea.

New API endpoint
`POST /doc/:docid/copy/:folderid`
By the way, I add the HTTP method.

docid is original document.
folderid is folder where copied document is saved.

FYI:
Document move API is /doc/:doc/move/:targetfolderid.
The suggested API is modified move API.

How about this, @0eta0.

**Reply by 0eta0** [reply]
I think this is more natural and better than the previous one!

**Comment by wancom** [comment]
I'll implement new API as follows.

New API endpoint
- `POST /doc/:docid/copy/:folderid`
By the way, I add the HTTP method.

`docid` is original document.
`folderid` is folder where copied document is saved.

FYI:
Document move API is `/doc/:doc/move/:targetfolderid`.
The suggested API is modified move API.

https://github.com/wonder-wonder/cakemix-server/discussions/232#discussioncomment-941374

**Comment by wancom** [comment]
The document duplication is implemented and merged by #242
